shell/src/main/java/alluxio/cli/fs/command/DistributedLoadCommand.java

```diff
@@ -130,29 +173,45 @@ public int run(CommandLine cl) throws AlluxioException, IOException {
    * @param filePath The {@link AlluxioURI} path to load into Alluxio memory
    * @param replication The replication of file to load into Alluxio memory
    */
-  private Callable<AlluxioURI> newJob(AlluxioURI filePath, int replication) {
-    return new Callable<AlluxioURI>() {
-      @Override
-      public AlluxioURI call() throws Exception {
-        JobGrpcClientUtils.run(new LoadConfig(filePath.getPath(), replication), 3,
-            mFsContext.getPathConf(filePath));
-        return filePath;
-      }
-    };
+  private JobAttempt newJob(AlluxioURI filePath, int replication) {
+    JobAttempt jobAttempt = new JobAttempt(new LoadConfig(filePath.getPath(), replication),
+        new CountingRetry(3), ClientContext.create(mFsContext.getPathConf(filePath)));
+
+    jobAttempt.run();
+
+    return jobAttempt;
   }
 
   /**
    * Waits one job to complete.
    */
-  private void waitJob() throws ExecutionException, InterruptedException {
+  private void waitJob() throws IOException {
+    boolean removed = false;
     while (true) {
-      Future<AlluxioURI> future = null;
-      // Take one completed job.
-      future = mDistributedLoadService.take();
-      if (future != null) {
-        AlluxioURI uri = future.get();
-        System.out.println(uri + " is loaded");
-        mFutures.remove(future);
+      Iterator<JobAttempt> iterator = mSubmittedJobAttempts.iterator();
+
+      while (iterator.hasNext()) {
+        JobAttempt jobAttempt = iterator.next();
+        Status check = jobAttempt.check();
+
+        switch (check) {
+          case CREATED:
+          case RUNNING:
+            continue;
+          case CANCELED:
+          case COMPLETED:
+            removed = true;
+            jobAttempt.close();
+            iterator.remove();
+            continue;
+          case FAILED:
+            jobAttempt.run();
```

Comment:
@bradyoo should we check whether the return value of `run` here? Otherwise, if the job keeps failing, the job cannot be removed from the queue, even when `run` returns `false`?

Reply:
You are right. I should fix this.